plat/common/include/uk/plat/common/lcpu.h

```diff
@@ -0,0 +1,367 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __PLAT_CMN_LCPU_H__
+#define __PLAT_CMN_LCPU_H__
+
+#include <uk/config.h>
+#include <uk/essentials.h>
+#ifndef __ASSEMBLY__
+#include <uk/arch/types.h>
+#include <uk/plat/lcpu.h>
+#include <uk/plat/spinlock.h>
+#include <uk/list.h>
+#endif /* !__ASSEMBLY__ */
+
+#if defined(__X86_64__)
+#include <x86/lcpu_defs.h>
+#else
+#error "Add lcpu_defs.h for current architecture."
+#endif
+
+/*
+ * LCPU Startup Arguments
+ */
+#define LCPU_SARGS_ENTRY_OFFSET		0x00
+#define LCPU_SARGS_STACKP_OFFSET	(LCPU_SARGS_ENTRY_OFFSET + 0x08)
+
+#define LCPU_SARGS_SIZE			0x10
+
+#ifndef __ASSEMBLY__
+struct lcpu_sargs {
+	ukplat_lcpu_entry_t entry;
+	void *stackp;
+};
+
+UK_CTASSERT(__offsetof(struct lcpu_sargs, entry)  == LCPU_SARGS_ENTRY_OFFSET);
+UK_CTASSERT(__offsetof(struct lcpu_sargs, stackp) == LCPU_SARGS_STACKP_OFFSET);
+
+UK_CTASSERT(sizeof(struct lcpu_sargs) == LCPU_SARGS_SIZE);
+#endif /* !__ASSEMBLY__ */
+
+/*
+ * Logical CPU (LCPU) Structure
+ */
+#define LCPU_STATE_OFFSET		0x00
+#define LCPU_IDX_OFFSET			(LCPU_STATE_OFFSET + 0x04)
+#define LCPU_ID_OFFSET			(LCPU_IDX_OFFSET   + 0x04)
+#define LCPU_ENTRY_OFFSET		(LCPU_ID_OFFSET    + 0x08)
+#define LCPU_STACKP_OFFSET		(LCPU_ENTRY_OFFSET + 0x08)
+#define LCPU_ERR_OFFSET			(LCPU_ENTRY_OFFSET + 0x00)
+#define LCPU_ARCH_OFFSET		(LCPU_ENTRY_OFFSET + 0x10)
+
+#define LCPU_MEMBERS_SIZE		(LCPU_ARCH_OFFSET   + LCPU_ARCH_SIZE)
+#define LCPU_SIZE						\
+	ALIGN_UP(LCPU_MEMBERS_SIZE, CACHE_LINE_SIZE)
+
+#ifndef __ASSEMBLY__
+struct lcpu {
+	/* Current CPU state (LCPU_STATE_*).
+	 * Working on it with atomic instructions - must be 8-byte aligned
+	 */
+	volatile int state __align(8);
+
+	__lcpuidx idx;
+	__lcpuid id;
+
+	union {
+		/* Startup arguments
+		 * Only valid in LCPU_STATE_INIT
+		 */
+		struct lcpu_sargs s_args;
+
+		/* Remote function to execute
+		 * Only valid in LCPU_STATE_IDLE and busy states
+		 */
+#ifdef CONFIG_HAVE_SMP
+		struct ukplat_lcpu_func fn;
+#endif /* CONFIG_HAVE_SMP */
+
+		/* Error code indicating the halt reason
+		 * Only valid in LCPU_STATE_HALTED
+		 */
+		int error_code;
+	};
+
+	/* Architectural-dependent part */
+	struct lcpu_arch arch;
+
+	/* We want every CPU on a separate cache line */
+	__u8 padding[LCPU_SIZE - LCPU_MEMBERS_SIZE];
+};
+
+UK_CTASSERT(sizeof(struct ukplat_lcpu_func)        == 0x10);
```

Comment:
Maybe replace 0x10 with a `#define` to be consistent with the check for `lcpu_sargs` at line 73